[P2][Checkup][Ability] Sturdy

flx-sta opened the pull request:
## What are the changes the user will see?

- The `Sturdy` ability will now trigger on self-inflicted confusion damage (given that it would go from 100% HP to 0% HP)

## Why am I making these changes?

- Closes #193 
- **[P2]** Closes #523 (Ability flyout issue will still be present!)
- **[P2]** Fixes: Sturdy now procs on self-inflicted dmg (e.g. confusion)

## What are the changes from a developer perspective?

- Sturdy ability test coverage enhancement.
- Add `SACRIFICIAL_MOVES`move constants.
- Add `toHaveHp` matcher
- Add `toHaveFainted` matcher
- Add `applyAbAttrs<PreDefendFullHpEndureAbAttr>(AbAttrFlag.PRE_DEFEND_FULL_HP_ENDURE, /*...*/);` call to `ConfusedTag.lapse` call

## Screenshots/Videos


<details><summary>[🎥 Boss enduring]</summary>
<p>

https://github.com/user-attachments/assets/1e651fe6-a9b4-45a8-80cf-c1c3230297bc

</p>
</details> 

<details><summary>[🎥 Enduring confusion self-damage]</summary>
<p>

https://github.com/user-attachments/assets/090b6231-c3ae-430f-a9d8-8631d37e1bea

</p>
</details> 


## How to test the changes?

`pnpm test:silent`

## Checklist

- [x] Otherwise: **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`pnpm test:silent`)
  - [x] Have I created new automated tests (`pnpm test:create`) or updated existing tests related to the PR's changes?

flx-sta committed:
add two custom matchers

- toHaveHp matcher
- toHaveFainted matcher

flx-sta committed:
Update docs related to "Sturdy" ability

flx-sta committed:
fix comment in confused-tag

flx-sta committed:
chore: ConfusedTag cleanup

flx-sta committed:
add SACRIFICIAL_MOVES constants

in "move-constants"

flx-sta committed:
add "PostKnockoutPhase" to PhaseInterceptor

flx-sta committed:
update Sturdy test

- bring "legacy" test cases up to date
- add more test coverage

flx-sta committed:
chore: ConfusedTag docs cleanup

flx-sta committed:
fix sturdy triggering on self-inflicted confusion damage

flx-sta committed:
activate study test for self-inflicted

flx-sta committed:
use `this.sourceMoveId` in `ConfusedTag.lapse`

flx-sta committed:
add ConfusedTag comment

flx-sta commented:
~~Taking a look at #523 too~~

Exceeds the range of a Checkup PR

Tempo-anon commented on test/abilities/sturdy.test.ts:
```diff
@@ -22,55 +44,224 @@ describe("Abilities - Sturdy", () => {
+    expect(enemy).toHaveFainted();
+  });
+
+  it.each(sacrificialMoves)("should not proc on sacrificial '%s' move", async (_enemyMoveName, enemyMoveId) => {
```
nit: should clarify this is user using a sacrificial move (or is the term self KO?)

flx-sta replied:
I _stole_ the term `sacrificial` from the Attr name. I thought it pretty clear

Tempo-anon commented on test/abilities/sturdy.test.ts:
```diff
@@ -22,55 +44,224 @@ describe("Abilities - Sturdy", () => {
-    expect(enemyPokemon.isFainted()).toBe(true);
+    expect(enemy).toHaveFainted();
   });
 });
```
wonder if there should be tests
* When it comes to bosses
* When it comes to status damage (though with the way the game works that only applies to Shedinja since that's the only time % damage = entire hp)
* For not interacting with wonder guard

flx-sta replied:
There should probably be tests for that, Good point.
It's possible that my boss testing will be lackluster though, as there seems to be quite some bugs involved in that

Tempo-anon commented on src/data/abilities/ab-attrs/pre-defend-full-hp-endure-ab-attr.ts:
```diff
@@ -5,6 +5,10 @@ import type { Pokemon } from "#field/pokemon";
+ * If the pokemon with this ab-attr is full hp and hit with a move that would kill it, it will survive with 1 hp left.
+ * @see {@link https://bulbapedia.bulbagarden.net/wiki/Sturdy_(Ability) | Sturdy Ability - Bulbapedia}
+ */
 export class PreDefendFullHpEndureAbAttr extends PreDefendAbAttr {
```
wonder if it makes sense to just name this SturdyAbAttr or something 🤔

flx-sta replied:
Sure, why not. I'll just double check it's usages

Tempo-anon commented on src/data/abilities/ab-attrs/pre-defend-full-hp-endure-ab-attr.ts:
```diff
@@ -5,6 +5,10 @@ import type { Pokemon } from "#field/pokemon";
 import type { NumberHolder } from "#utils/common-utils";
 
+/**
+ * If the pokemon with this ab-attr is full hp and hit with a move that would kill it, it will survive with 1 hp left.
```
I think it will be beneficial to add additional information here like how it interacts with boss bars (no mainline correlation) and sturdy (does not work unlike mainline)

Tempo-anon reviewed (dismissed):
While the PR does a checkup, I think it's worth adding in the actual change in the PR title as well (sturdy now procs on self confusion hit)

flx-sta committed:
update PreDefendFullHpEndureAbAttr docs

add details about Boss Pokemon

flx-sta committed:
rename PreDefendFullHpEndureAbAttr to SturdyAbAttr

flx-sta committed:
update Sturdy test

- add "self-KO" info to sacrificial moves

flx-sta committed:
Merge branch 'beta' into 193-abilitycheckup-sturdy

flx-sta committed:
update SturdyAbAttr docs

- add info about wonder guard

flx-sta committed:
add docs to EnemyPokemon fields

flx-sta committed:
extend Sturdy tests

- add tests for boss (but put into "todo" with info about the ongoing bug)
- add test for wonder guard
- add tests for status damage with 1 max hp
- add tests for damaging trapped battler tag types with 1 max hp

flx-sta committed:
extend Sturdy tests

- add seeded, nightmare, cursed and salt-cured to the damaging battler-tag types

DayKev commented on src/data/abilities/ab-attrs/sturdy-ab-attr.ts:
```diff
@@ -5,7 +5,20 @@ import type { Pokemon } from "#field/pokemon";
+ * If the pokemon with this ab-attr is full hp and hit with a move that would 1-hit faint it, it will survive with 1 hp left.
+ * Wonder Guard _overrules_ this ab-attr.
+ *
+ * #### Boss Pokemon
```
Is this just indirect documentation of the bug?
The ability prevents fainting if the user would go from full HP to 0 HP in one hit, causing the user to survive with 1 HP left. Boss bars make that more difficult (basically increasing the pokemon's effective HP, but not actual HP), but shouldn't change how the ability activates.

flx-sta replied:
This isn't documenting the bug, it's documenting the actual expected behavior with an example of how it would be expected to 1hit it.
If it's not clear enough let me know. It's rather complicated thus hard to summarize in few words without going too much into the weeds

DayKev replied:
I don't think it makes sense to document how boss bars work here, they should be documented at the fields/methods/etc related to them. You don't need to know about boss bars to understand Sturdy because there's no interaction between them.

flx-sta replied:
But this is specific to the Sturdy behavior. It's not universal boss explanation but how sturdy interacts with boss segments. It's unique to PKTY

DayKev replied:
Not really? You're just explaining how much damage is needed to KO a Pokémon with boss bars. Sturdy cares if a Pokémon goes from full HP to 0 HP; yes it takes more damage to do it when the Pokémon has boss bars because they "absorb" some of the damage but there's no actual interaction between Sturdy and the boss bars.
Sturdy sees a Pokémon going from full HP to 0 HP and sets it to 1 HP, it doesn't see the boss bars.
Boss bars are just one of many methods to effectively reduce the damage a Pokémon is taking.

flx-sta replied:
@DayKev Where should i put this explanation/documentation instead then?

DayKev replied:
Somewhere boss bars are defined/handled (maybe in `EnemyPokemon#damage`, at L544?).
https://github.com/Despair-Games/poketernity/blob/ee5269b594e7deafc4cb68a0e3f08a169eba4a61/src/field/enemy-pokemon.ts#L544-L546

DayKev commented on src/data/battler-tags/confused-tag.ts:
```diff
@@ -60,39 +66,56 @@ export class ConfusedTag extends BattlerTag {
-        pokemon.damageAndUpdate(damage);
+
+        if (pokemon.isFullHp()) {
+          // `applyAbAttrs` may modify `damageHolder` by reference to adjust the damage value.
```
```suggestion
```
These would just be nuisance comments if we had them before every single `applyAbAttrs` call.

Edit: Also this is just inaccurate anyway, the `AbAttr` doesn't modify the damage dealt, it just applies a `BattlerTag` that handles the actual damage modification.

DayKev commented on src/data/battler-tags/confused-tag.ts:
```diff
@@ -60,39 +66,56 @@ export class ConfusedTag extends BattlerTag {
+        if (pokemon.isFullHp()) {
+          // `applyAbAttrs` may modify `damageHolder` by reference to adjust the damage value.
+          applyAbAttrs<SturdyAbAttr>(
+            AbAttrFlag.PRE_DEFEND_FULL_HP_ENDURE,
```
Since you renamed the `AbAttr` you should rename the matching `AbAttrFlag` too.

DayKev commented on src/data/battler-tags/sturdy-tag.ts:
```diff
@@ -10,8 +10,7 @@ import i18next from "i18next";
+ * This is added to Pokemon with the ability whenever they receive lethal attack damage from full HP.
  * @extends BattlerTag
  */
 export class SturdyTag extends BattlerTag {
```
~Hm, is this `BattlerTag` really needed for the message to appear at the right time? It seems to me we should just be able to use `appendToPhase(messagePhase, PhaseId.MOVE_EFFECT)`.~
~https://www.youtube.com/watch?v=dEUO7E_ozQA~

Edit: This tag seems to be what's applying the actual effects of Sturdy, meaning this TSDoc is inaccurate (it doesn't just handle the messages).

DayKev commented on test/abilities/sturdy.test.ts:
```diff
@@ -25,50 +63,312 @@ describe("Abilities - Sturdy", () => {
+
+  it("should not proc on 'Destiny Bond' move", async () => {
+    const { classicMode, field, move } = game;
+    vi.spyOn(DestinyBondTag.prototype, "lapse");
```
You never checked that these functions weren't called (I assume that's the intention of spying on them?).

DayKev commented on test/abilities/sturdy.test.ts:
```diff
@@ -25,50 +63,312 @@ describe("Abilities - Sturdy", () => {
-    await game.classicMode.startBattle(SpeciesId.LUCARIO);
-    game.move.select(MoveId.FISSURE);
-    await game.phaseInterceptor.to("PostActionPhase");
+  it("should not proc when 'Wonder Guard' ability is present too", async () => {
```
Is this just testing that the flyout doesn't show incorrectly?

There should be a test showing that Sturdy doesn't activate if a Wonder Guard pokemon takes damage from an attack (it should faint).

DayKev commented on test/test-utils/matchers/to-have-fainted-matcher.ts:
```diff
@@ -0,0 +1,31 @@
+
+  const { hp } = received;
+  const maxHp = received.getMaxHp();
+  const pass = received.isFainted() === true && hp === 0;
```
```suggestion
  const pass = received.isFainted() === true && hp === 0;
```
`Pokemon#isFainted` is already checking for the pokemon having 0 HP.

DayKev commented:
Commented on #523 with info on fixing the bugs from it. The functionality bug should be a simple fix, the ability flyout is more complicated but also just a P3 so 🤷.

flx-sta committed:
Update src/data/battler-tags/confused-tag.ts

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>

flx-sta committed:
Merge branch 'beta' into 193-abilitycheckup-sturdy

flx-sta committed:
update toHaveFaintedMatcher

- remove redundancy of "pass" check

flx-sta committed:
rename `AbAttrFlag.  PRE_DEFEND_FULL_HP_ENDURE`

- to `AbAttrFlag.STURDY`

flx-sta committed:
update SturdyTag doc

flx-sta committed:
Update Sturdy test

- fix Destiny Bond proc check
- fix Wonder Guard proc check

flx-sta committed:
Merge branch 'beta' into 193-abilitycheckup-sturdy

flx-sta committed:
fix #523

flx-sta committed:
Update Sturdy Test

- activate the boss test (and fix it)

flx-sta commented:
> Commented on #523 with info on fixing the bugs from it. The functionality bug should be a simple fix, the ability flyout is more complicated but also just a P3 so 🤷.

Added the fix for that to the PR

DayKev commented on src/data/battler-tags/sturdy-tag.ts:
```diff
@@ -8,10 +8,9 @@ import type { Pokemon } from "#field/pokemon";
 
 /**
- * Tag to manage messages related to the enduring effect of
+ * Tag to behavior related to the enduring effect of
```
```suggestion
 * Tag to implement behavior related to the enduring effect of
```

DayKev commented on test/abilities/sturdy.test.ts:
```diff
@@ -25,50 +62,311 @@ describe("Abilities - Sturdy", () => {
-  test("Sturdy doesn't activate when user is not at full HP", async () => {
-    await game.classicMode.startBattle(SpeciesId.LUCARIO);
+  it("should proc properly on Boss Pokemon and deplete all hp-segments but the last", async () => {
+    // Known issue: Boss Pokemon sturdy triggers at the end of the first hp-segment, instead of the last (See #523)
```
```suggestion
```

DayKev replied:
```suggestion
    // Known issue: Boss Pokemon sturdy flyout triggers at the end of the first hp-segment, instead of the last (See #523)
```

flx-sta committed:
update SturdyTag docs

flx-sta committed:
Update comment in Sturdy test

flx-sta committed:
Merge branch 'beta' into 193-abilitycheckup-sturdy

flx-sta committed:
rename `toHaveBattleTagTypeMatcher`  to `toHaveBattlerTagMatcher`

flx-sta committed:
Merge branch 'beta' into 193-abilitycheckup-sturdy

flx-sta committed:
move boss docs into enemy-pokemon.ts

flx-sta committed:
Merge branch 'beta' into 193-abilitycheckup-sturdy

flx-sta committed:
Merge branch 'beta' into 193-abilitycheckup-sturdy

flx-sta committed:
fix tests

- replace "toHaveBattlerTagType" with "toHaveBattlerTag"

innerthunder commented on src/constants/move-constants.ts:
```diff
@@ -5,3 +5,12 @@ export const TERA_MOVES = Object.freeze([MoveId.TERA_BLAST, MoveId.TERA_STARSTOR
 export const PLEDGE_MOVES = Object.freeze([MoveId.GRASS_PLEDGE, MoveId.FIRE_PLEDGE, MoveId.WATER_PLEDGE]);
+
+/** An array of all sacrificial {@linkcode MoveId | MoveIds}. */
+export const SACRIFICIAL_MOVES = Object.freeze([
```
Should this also include Healing Wish and Lunar Dance?
<sub>Given they require a non-fainted Pokemon in the user's party, you'd probably have to modify tests a bit if they're added</sub>

flx-sta replied:
Probably, yeah.
https://pokemon.fandom.com/wiki/Category:Sacrificing_Moves

innerthunder commented on src/data/abilities/ab-attrs/sturdy-ab-attr.ts:
```diff
@@ -5,10 +5,16 @@ import type { Pokemon } from "#field/pokemon";
 
-export class PreDefendFullHpEndureAbAttr extends PreDefendAbAttr {
+/**
+ * If the pokemon with this ab-attr is full hp and hit with a move that would 1-hit faint it, it will survive with 1 hp left.
+ * Wonder Guard _overrules_ this ab-attr.
```
```suggestion
 * If the pokemon with this ab-attr is full hp and hit with a move that would 1-hit faint it, it will survive with 1 hp left (_unless it also has the ability Wonder Guard_).
```
It's not immediately clear what "overruling" the attribute entails. You may want to break this up to reduce line length as well.

DayKev replied:
```suggestion
 * If the pokemon with this `AbAttr` is full HP and is hit with a move that would 1-hit faint it,
 * it will survive with 1 HP left (_unless it also has the ability Wonder Guard_).
```
Modified inner's suggestion to break it into 2 lines.

DayKev commented on test/abilities/sturdy.test.ts:
```diff
@@ -25,50 +62,311 @@ describe("Abilities - Sturdy", () => {
+    const { override, classicMode, field, move } = game;
+    override
+      .enemySpecies(SpeciesId.SHEDINJA)
+      .enemyPassiveAbility(AbilityId.WONDER_GUARD)
```
Wonder Guard already prevents Sturdy from activating, so this test isn't actually checking that status damage doesn't proc Sturdy.
I think it should be impossible for both Burn and Poison to OHKO a Pokémon anyway since they deal %-based damage, and after checking Bulbapedia Toxic apparently can only do a max of `15/16` of a Pokémon's max HP (which I don't think we do correctly actually lol).
So in the end, these status tests are unnecessary.

flx-sta replied:
Not sure why I used Wonder-Guard here in the first place tbh.
Also not sure if I should keep this test at all if I were to remove it as it would just be redundant to  `"should not proc on '%s' damaging trap battler-tag ",`

DayKev commented on test/abilities/sturdy.test.ts:
```diff
@@ -25,50 +62,311 @@ describe("Abilities - Sturdy", () => {
+    expect(enemy).toHaveFainted();
   });
+
+  it.each(damagingBattlerTagTypes)(
```
These moves/effects do %-based damage and so should never be able to OHKO a Pokémon. If they did, that would be a bug.

flx-sta replied:
But is it wrong to cover it in a test? I mean, look at MEs, who knows what other ideas might be added in the future

DayKev replied:
If we introduce a new effect like that then a test for that effect should be made, in which case these tests would still be redundant.

DayKev replied:
These tests do have a time cost (~450ms per test for me). If we're testing something relevant that's fine, we shouldn't worry about it. These tests aren't testing for anything that's possible in the game though.

Also you duplicated most of the `BattlerTag` tests as well.
![image](https://github.com/user-attachments/assets/99f018cd-4202-4e9d-bea0-0cbd0b2baced)

flx-sta committed:
Update src/data/abilities/ab-attrs/sturdy-ab-attr.ts

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>

flx-sta commented on test/abilities/sturdy.test.ts:
```diff
@@ -25,50 +62,311 @@ describe("Abilities - Sturdy", () => {
+      await game.toEndOfTurn();
+
+      expect(SturdyAbAttr.prototype.apply).not.toHaveBeenCalled();
+      expect(TrappedTag.prototype.lapse).toHaveBeenCalled;
```
```suggestion
      expect(TrappedTag.prototype.lapse).toHaveBeenCalled();
```

flx-sta committed:
Update test/abilities/sturdy.test.ts

flx-sta committed:
Merge branch 'beta' into 193-abilitycheckup-sturdy

flx-sta committed:
update Sturdy tests

flx-sta committed:
remove tests from Sturdy test

flx-sta committed:
add healing-wish and lunar-dance to SACRIFICIAL_MOVES

flx-sta committed:
update Sturdy test for sacrificial moves

- make sure that lunar dance & healing wish are tested properly too by turning around the "perspective"

flx-sta committed:
Merge branch 'beta' into 193-abilitycheckup-sturdy

flx-sta committed:
update Sturdy test

- remove "damaging battle tag trap" test
- remove "Focus sash" todo test

DayKev committed:
Merge branch 'beta' into 193-abilitycheckup-sturdy
